src/snap.py

```diff
@@ -0,0 +1,153 @@
+        return snap_lib.SnapCache()[_SNAP_NAME]
+
+    def install(self, *, unit: ops.Unit):
+        """Install snap."""
+        if self._snap.present:
+            logger.error(f"{_SNAP_NAME} snap already installed on machine. Installation aborted")
+            raise Exception(f"Multiple {_SNAP_NAME} snap installs not supported on one machine")
+        logger.debug(f"Installing {_SNAP_NAME=}, {self._SNAP_REVISION=}")
+        unit.status = ops.MaintenanceStatus("Installing snap")
+
+        def _set_retry_status(_) -> None:
+            unit.status = ops.MaintenanceStatus("Snap install failed. Retrying...")
+
+        try:
+            for attempt in tenacity.Retrying(
+                stop=tenacity.stop_after_delay(60 * 5),
+                wait=tenacity.wait_exponential(multiplier=10),
+                retry=tenacity.retry_if_exception_type(snap_lib.SnapError),
+                after=_set_retry_status,
+                reraise=True,
+            ):
+                with attempt:
+                    self._snap.ensure(
+                        state=snap_lib.SnapState.Present, revision=self._SNAP_REVISION
+                    )
+        except snap_lib.SnapError:
+            raise
+        logger.debug(f"Installed {_SNAP_NAME=}, {self._SNAP_REVISION=}")
+
+    def uninstall(self):
+        """Uninstall snap."""
+        logger.debug(f"Uninstalling {_SNAP_NAME=}")
+        self._snap.ensure(state=snap_lib.SnapState.Absent)
+        logger.debug(f"Uninstalled {_SNAP_NAME=}")
+
+
+class _Path(pathlib.PosixPath, container.Path):
+    def __new__(cls, *args, **kwargs):
+        path = super().__new__(cls, *args, **kwargs)
+        if str(path).startswith("/etc/mysqlrouter") or str(path).startswith(
+            "/var/lib/mysqlrouter"
+        ):
+            parent = f"/var/snap/{_SNAP_NAME}/current"
+        elif str(path).startswith("/run"):
+            parent = f"/var/snap/{_SNAP_NAME}/common"
+        elif str(path).startswith("/tmp"):
+            parent = f"/tmp/snap-private-tmp/snap.{_SNAP_NAME}"
+        else:
+            parent = None
+        if parent:
+            assert str(path).startswith("/")
+            path = super().__new__(cls, parent, path.relative_to("/"), **kwargs)
+        path._container_parent = parent
+        return path
+
+    def __rtruediv__(self, other):
+        return type(self)(other, self)
+
+    @property
+    def relative_to_container(self) -> pathlib.PurePosixPath:
+        if parent := self._container_parent:
+            return pathlib.PurePosixPath("/", self.relative_to(parent))
+        return self
+
+    def read_text(self, encoding="utf-8", *args) -> str:
+        return super().read_text(encoding, *args)
+
+    def write_text(self, data: str, encoding="utf-8", *args):
+        return super().write_text(data, encoding, *args)
+
+    def rmtree(self):
+        shutil.rmtree(self)
+
+
+class Snap(container.Container):
+    """Workload snap container"""
+
+    _SNAP_REVISION = "57"
```

Comment:
`_SNAP_REVISION` is present here and on `Installer`
Why not merge to a single module literal? Or why not ditch from `Snap` since it's not being referenced?